lsp-mode.el

```diff
@@ -4230,7 +4238,7 @@ The method uses `replace-buffer-contents'."
                            (-when-let ((&SnippetTextEdit :range (&RangeToPoint :start)
                                                          :insert-text-format? :new-text) edit)
                              (when (eq insert-text-format? lsp/insert-text-format-snippet)
-                               (let (yas-indent-line)
+                               (let ((yas-indent-line (lsp--indent-snippets?)))
```

Comment:
Instead of that, why don't we use something like 
```suggestion
                               (let ((yas-indent-line 'auto)
                                     (indent-line-function 'indent-relative-first-indent-point)))
```

Since I believe the server should return the snippet with some indentation already (according to server format), we shouldn't apply client indentation logic again.
However, we should insert server snippet relative to current indent point

Reply:
> Since I believe the server should return the snippet with some indentation already (according to server format)

This, unfortunately, is not true - the indentation should be performed client side.

> (indent-line-function 'indent-relative-first-indent-point)

I tested it and it does not yield good results since everything is aligned to the first point which is incorrect.  Check the example at https://github.com/emacs-lsp/lsp-mode/issues/1805

Reply:
> > Since I believe the server should return the snippet with some indentation already (according to server format)
> 
> This, unfortunately, is not true - the indentation should be performed client side.

Given that the server side format and client side format can be different and `lsp-format-buffer` will format the current buffer according to server side format, I think that client side should not perform indentation according to major mode.
Even with #1805, the server is returning snippet with correct indentation, just that the insert point in client side has some offset, and those offsets are not respected when we expand the snippet

> I tested it and it does not yield good results since everything is aligned to the first point which is incorrect.

Then we may need to find another function so that the indentation is simply shift everything according to where the first line is expanded.


> Check the example at #1805

The problem in #1805 is because the snippet is not indented, you can see that the first line in the snippet expansion is inserted with some offset from line begin, whereas the rest is inserted without offset.
If those following line in snippet is inserted with offset, then we will have working snippet with correct indentation.

Reply:
Consider the following case:
```
    /** 123 */
    public static void main(final String[] args) 
```

1/2/3 are the positions in which you invoke the completion. In the 3 cases the indentation returned by the server is `"\n * ${0}\n * @param args\n"` and it the 3 cases it is expected to end up with the same result so the indentation is not relative to the point in which the completion has been requested.

Reply:
I found this LSP thread
https://github.com/microsoft/language-server-protocol/issues/83
And also some proposed from VsCode lsp extension
https://github.com/microsoft/vscode/issues/57093
https://github.com/Microsoft/vscode/blob/25589f6cc4aa3450450fa0f49979fcf7835f52d4/src/vs/vscode.proposed.d.ts#L16-L39

Without looking at the real implementation, from the proposed extension, it leaves an impression on me that VsCode is actually (proposing to other editor) to just adjust leading whitespace of new lines to match with the indentation of the line where snippet is inserted.
``` ts
/**
 * Keep whitespace as is. By default, the editor adjusts leading
 * whitespace of new lines so that they match the indentation of
 * the line for which the item is accepted.
 */
```

Also, if we turn on magic indentation by client, I doubt `yaml-mode` may have been broken by that, since spaces has special meaning for it.

Come back to your case, since the snippet is inserted in the first line, the consequence new lines should be insert at the offset of the first line indentation, which works perfectly here.
And also that will not slow down `org-mode`.
We will need special handling for `yaml-mode` though, may be put that as a field (opt-out) in `lsp-client`?

Reply:
> Also, if we turn on magic indentation by client, I doubt `yaml-mode` may have been broken by that, since spaces has special meaning for it.

We were running years with indenting being on until I stopped it when I was doing `org-mode` integration a few months ago so most likely there won't be any issues with having that on.

Is there a case that is unhandled by this PR?

Reply:
Ok, I didn't notice that. Nevertheless I think doing indent based on the line where snippet is inserted is better and not burden Emacs side. Thus it can even apply to `org-mode` and will ensure same kind of formatting between server and client

Reply:
Those can be address in another PR though